src/mgr/ActivePyModules.cc

```diff
@@ -163,8 +165,53 @@ PyObject *ActivePyModules::get_daemon_status_python(
   return f.get();
 }
 
+bool inject_python_on() {
+  return g_conf().get_val<bool>("mgr_inject");
+}
+
+bool ttl_cache_on() {
+  return g_conf().get_val<uint64_t>("mgr_ttl_cache_expire_seconds") > 0;
+}
+
+void ActivePyModules::update_ttl_cache() {
+  ttl_cache.set_ttl(g_conf().get_val<uint64_t>("mgr_ttl_cache_expire_seconds"));
+}
+
 PyObject *ActivePyModules::get_python(const std::string &what)
 {
+  update_ttl_cache();
+  if(ttl_cache_on()) {
+    boost::optional<PyObject*> cached = ttl_cache.get(what);
+    ttl_cache.set_erase_callback([](PyObject *obj) {
+	Py_DECREF(obj);
+    });
```

Comment:
Ok, got it! In this very case, rather than the explicit callback, I'd suggest to encapsulate this by writing a specialization for the class template for `Value = PyObject*`. Any cache dealing with PyObjects will have to take care of the ref counting.

Additionally, I was going to suggest using an smart pointer (like `shared_ptr`), but in this very case you'd need a smart pointer that delegates the ref-counting, so what about [`boost::intrusive_ptr`](https://www.boost.org/doc/libs/1_60_0/libs/smart_ptr/intrusive_ptr.html)? You basically connect it with the [Py_INCREF and Py_DECREF functions](https://codereview.stackexchange.com/a/120550):
```c++
    void intrusive_ptr_add_ref(PyObject *p)
    {
        Py_INCREF(p);
    }

    void intrusive_ptr_release(PyObject *p)
    {
        Py_DECREF(p);
    }
```

Other approach is that you go with either a smart cache (it calls the `get("osd_map")`, stores the retval, and releases it when expiring) or a dumb cache (it doesn't take care of what's holding in it and doesn't deallocate anything).